File: app/src/main/java/com/example/dices2/MainActivity.java
package com.example.dices2;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.TableLayout;

import java.util.List;

import static com.example.dices2.Consts.*;

public class MainActivity extends AppCompatActivity {
    private Game game;
    private TableLayout tableLayoutMain;
    private GameTable gameTable;
    private final int MENU_EDIT = 1;

    // TODO поиграться с цветами

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        init();
        game = Game.getInstance();
        game.setMainActivity(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.new_game:
                if (game.isGameStarted()) {
                    AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(MainActivity.this);
                    alertDialogBuilder.setTitle("Прервать игру");
                    alertDialogBuilder.setMessage("Вы уверены, что хотите прервать текущую игру?");
                    alertDialogBuilder.setNegativeButton("Нет", null);
                    alertDialogBuilder.setPositiveButton("Да", new AlertDialog.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            startListActivity();
                        }
                    });
                    alertDialogBuilder.show();
                } else {
                    startListActivity();
                }
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void startListActivity() {
        Intent intent = new Intent(this, ListActivity.class);
        startActivity(intent);
    }

    @Override
    protected void onActivityResult (int requestCode, int resultCode, Intent data ) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data == null) return;
        String value = data.getStringExtra(INTENT_INPUT_VALUE);
        boolean isEdit = data.getBooleanExtra(INTENT_IS_EDIT, false);
        game.okButtonOnKeyboardActivityClicked(value, isEdit);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if(keyCode == KeyEvent.KEYCODE_BACK)
        {
            AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(MainActivity.this);
            alertDialogBuilder.setTitle("Выход");
            alertDialogBuilder.setMessage("Вы уверены, что хотите выйти?");
            alertDialogBuilder.setNegativeButton("Нет", null);
            alertDialogBuilder.setPositiveButton("Да", new AlertDialog.OnClickListener() {
                public void onClick(DialogInterface dialog, int which) {
                    finish();
                }});
            alertDialogBuilder.show();
        }
        return false;
    }

    public void fillMainTable(List<Player> players) {
        gameTable = new GameTable(this, tableLayoutMain);
        gameTable.fillTable(players);
    }

    public void showSchool(Player player) {
        gameTable.setSchoolValue(player);
    }

    public void showTotal(Player player) {
        gameTable.setTotalValue(player);
    }

    private void init() {
        tableLayoutMain = findViewById(R.id.tl_main);
    }

    public void setTextViewHighlight(Cell currentCell, boolean isHighlight) {
        if (isHighlight) {
            currentCell.setBackground(getResources().getDrawable(R.drawable.text_view_back_light_gray));
        } else {
            currentCell.setBackground(getResources().getDrawable(R.drawable.text_view_back_dark_gray));
        }
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View view, ContextMenu.ContextMenuInfo info) {
        menu.add(0, MENU_EDIT, 0, "Редактировать");
        game.setCurrentCell((Cell)view);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        if (item.getItemId() == MENU_EDIT) {
            game.onEditMenuClicked();
        }
        return super.onContextItemSelected(item);
    }

    public void startKeyboardActivityForEdit(String editedValue, RowName rowName) {
        Intent intent = new Intent(this, KeyboardActivity.class);
        intent.putExtra(INTENT_EDITED_VALUE, editedValue);
        intent.putExtra(INTENT_ROW_NAME, rowName.getName());
        intent.putExtra(INTENT_IS_EDIT, true);
        startActivityForResult(intent, 1);
    }

    public void highlightPlayersMove(Player player) {
        gameTable.highlightPlayersMove(player);
    }

    public void switchOffPlayersMove(Player player) {
        gameTable.switchOffPlayersMove(player);
    }

    public void enableFullSquarePokerTextViews(Player player) {
        gameTable.enableAfterThreeClassesTextViews(player);
    }

    public void disableFullSquarePokerTextViews(Player player) {
        gameTable.disableAfterThreeClassesTextViews(player);
    }

    public void setClickedCell(Cell cell) {
        game.setCurrentCell(cell);
    }

    public void showPlayerPlace(Player player, int place) {
        gameTable.setPlayerPlace(player, place);
    }
}

File: app/src/main/java/com/example/dices2/MainActivity.kt
package com.example.dices2

import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Bundle
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import com.example.dices2.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var game: Game
    private lateinit var gameTable: GameTable
    private lateinit var pref: SharedPreferences

    // TODO поиграться с цветами

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        pref = getSharedPreferences(Consts.APP_PREFERENCES, MODE_PRIVATE)

        game = Game.instance
        game.setMainActivity(mainActivity = this)
    }

    override fun onResume() {
        super.onResume()

        if (!game.isGameStarted) {
            if (pref.contains(Consts.APP_PREFERENCES_TOTAL)) {
                game.isCountTotalEveryMove = pref.getBoolean(Consts.APP_PREFERENCES_TOTAL, false)
            }

            if (pref.contains(Consts.APP_PREFERENCES_PLACE)) {
                game.isCountPlaceEveryMove = pref.getBoolean(Consts.APP_PREFERENCES_PLACE, false)
            }

            if (pref.contains(Consts.APP_PREFERENCES_SCREEN)) {
                game.isKeelScreenOn = pref.getBoolean(Consts.APP_PREFERENCES_SCREEN, false)
                if (game.isKeelScreenOn) {
                    window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                } else {
                    window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                }
            }

            if (pref.contains(Consts.APP_PREFERENCES_PLAYERS)) {
                game.fillStartPlayers(playersSet = pref.getStringSet(Consts.APP_PREFERENCES_PLAYERS, HashSet<String>()))
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.new_game -> {
                if (game.isGameStarted) {
                    val alertDialogBuilder = AlertDialog.Builder(this@MainActivity)
                    alertDialogBuilder.setTitle(getString(R.string.break_game))
                    alertDialogBuilder.setMessage(getString(R.string.are_you_sure_to_break_game))
                    alertDialogBuilder.setNegativeButton(getString(R.string.no), null)
                    alertDialogBuilder.setPositiveButton(getString(R.string.yes)) { _, _ -> startListActivity() }
                    alertDialogBuilder.show()
                } else {
                    startListActivity()
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun startListActivity() {
        val intent = Intent(this, ListActivity::class.java)
        startActivity(intent)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data == null) return
        var value = ""
        data.getStringExtra(Consts.INTENT_INPUT_VALUE)?.let { value = it }
        val isEdit = data.getBooleanExtra(Consts.INTENT_IS_EDIT, false)
        game.okButtonOnKeyboardActivityClicked(value = value, isEdit = isEdit)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (game.isGameStarted) {

                val alertDialogBuilder = AlertDialog.Builder(this@MainActivity)
                alertDialogBuilder.setTitle(getString(R.string.exit))
                alertDialogBuilder.setMessage(getString(R.string.are_you_sure_to_exit))
                alertDialogBuilder.setNegativeButton(getString(R.string.no), null)
                alertDialogBuilder.setPositiveButton(getString(R.string.yes)) { _, _ -> exit() }
                //alertDialogBuilder.create()
                val dialog = alertDialogBuilder.create()
                dialog.show()
                dialog.getButton(Dialog.BUTTON_NEGATIVE).setTextColor(resources.getColor(R.color.no, resources.newTheme()))
                dialog.getButton(Dialog.BUTTON_POSITIVE).setTextColor(resources.getColor(R.color.yes, resources.newTheme()))
                dialog.getButton(Dialog.BUTTON_POSITIVE).textSize = 24f
            } else {
                exit()
            }
        }
        return false
    }

    private fun exit() {
        game.isGameStarted = false
        game.savePlayers(pref = pref)
        finish()
    }

    fun fillMainTable(players: List<Player>, player: Player?) {
        gameTable = GameTable(mainActivity = this, mainTable = binding.tlMain)
        gameTable.fillTable(players = players, player = player)
    }

    fun showSchool(player: Player) {
        gameTable.setSchoolValue(player = player)
    }

    fun showTotal(player: Player) {
        gameTable.setTotalValue(player = player)
    }

    fun setTextViewHighlight(currentCell: Cell, isHighlight: Boolean) {
        if (isHighlight) {
            // TODO getDrawable deprecated
            currentCell.background = resources.getDrawable(R.drawable.text_view_back_light_gray)
        } else {
            currentCell.background = resources.getDrawable(R.drawable.text_view_back_dark_gray)
        }
    }

    fun startKeyboardActivityForEdit(cell: Cell) {
        val intent = Intent(this, KeyboardActivity::class.java)
        intent.putExtra(Consts.INTENT_EDITED_VALUE, cell.value)
        intent.putExtra(Consts.INTENT_ROW_NAME, cell.row.getName())
        intent.putExtra(Consts.INTENT_IS_EDIT, true)
        intent.putExtra(Consts.INTENT_IS_IN_SCHOOL, gameTable.isCellInSchool(cell = cell))
        startActivityForResult(intent, 1)
    }

    fun highlightPlayersMove(player: Player?) {
        gameTable.highlightPlayersMove(player = player)
    }

    fun switchOffPlayersMove(player: Player) {
        gameTable.switchOffPlayersMove(player = player)
    }

    fun enableFullSquarePokerTextViews(player: Player) {
        gameTable.enableAfterThreeClassesTextViews(player = player)
    }

    fun disableFullSquarePokerTextViews(player: Player) {
        gameTable.disableAfterThreeClassesTextViews(player = player)
    }

    fun setClickedCell(cell: Cell) {
        game.setCurrentCell(cell = cell)
    }

    fun showPlayerPlace(player: Player) {
        gameTable.setPlayerPlace(player = player)
    }

    val isCountTotalEveryMove: Boolean
        get() = game.getIsCountTotalEveryMove()

    val isCountPlaceEveryMove: Boolean
        get() = game.getIsCountPlaceEveryMove()

}